zipline/pipeline/loaders/blaze/_core.pyx

```diff
@@ -0,0 +1,651 @@
+            )
+        else:
+            _ffill_missing_value_2d_inplace_impl[np.float64_t](
+                array,
+                missing_value,
+                is_missing_value[np.float64_t],
+            )
+    elif kind == 'O':
+        _ffill_missing_value_2d_inplace_impl[object](
+            array,
+            missing_value,
+            is_missing_value[object],
+        )
+    else:
+        raise TypeError('unknown column dtype: %r' % array.dtype)
+
+
+
+@cython.final
+cdef class AsAdjustedArray:
+    """Marker type for array_for_column which enables the AdjustedArray
+    path.
+    """
+
+
+@cython.final
+cdef class AsBaselineArray:
+    """Marker type for array_for_column which enables the baseline array
+    path.
+    """
+
+
+ctypedef fused AsArrayKind:
+    AsAdjustedArray
+    AsBaselineArray
+
+
+cdef _array_for_column_impl(object dtype,
+                            np.ndarray[column_type, ndim=2] out_array,
+                            Py_ssize_t size,
+                            np.ndarray[np.int64_t] ts_ixs,
+                            np.ndarray[np.int64_t] asof_ixs,
+                            np.ndarray[np.int64_t] sids,
+                            dict column_ixs,
+                            object mask,
+                            np.ndarray[column_type] input_array,
+                            column_type missing_value,
+                            bint is_missing(column_type, column_type),
+                            AsArrayKind _):
+    cdef column_type value
+    cdef np.int64_t ts_ix
+    cdef np.int64_t asof_ix
+    cdef np.int64_t sid
+    cdef np.int64_t column_ix
+
+    cdef PyObject* adjustments_list_ptr
+    cdef list adjustments_list
+
+    cdef list non_null_ixs
+    cdef dict non_null_ixs_by_sid = {sid: [] for sid in sids}
+    cdef dict adjustments
+
+    cdef Py_ssize_t out_of_bounds_ix = len(out_array)
+
+    if AsArrayKind is AsAdjustedArray:
+        adjustments = {}
+
+    cdef set categories
+    if column_type is object:
+        # for object columns we need to maintain the unique values for the
+        # categories
+        categories = set()
+
+    cdef Py_ssize_t n
+    for n in range(size if len(out_array) else 0):
+        with cython.boundscheck(False), cython.wraparound(False):
+            value = input_array[n]
+
+        if is_missing(value, missing_value):
+            # skip missing values
+            continue
+
+        if column_type is object:
+            # maintain the categories for the label array if we have an object
+            # column
+            categories.add(value)
+
+        with cython.boundscheck(False), cython.wraparound(False):
+            ts_ix = ts_ixs[n]
+            if ts_ix == out_of_bounds_ix:
+                # this timestamp falls after the last date requested
+                continue
+
+            asof_ix = asof_ixs[n]
+            if asof_ix == out_of_bounds_ix:
+                raise ValueError('asof_date newer than timestamp')
```

Comment:
if this happens at all it probably happens in a lot of rows. in practice when you hit this case you need to run through the debugger to look at the dataset and see what is broken